src/Compilers/Helpers/AbstractAnalyzerAssemblyLoader.cs

```diff
@@ -0,0 +1,186 @@
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Reflection.Metadata;
+using System.Reflection.PortableExecutable;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal abstract class AbstractAnalyzerAssemblyLoader : IAnalyzerAssemblyLoader
+    {
+        private readonly Dictionary<string, Assembly> _pathsToAssemblies = new Dictionary<string, Assembly>(StringComparer.OrdinalIgnoreCase);
+        private readonly Dictionary<string, Assembly> _namesToAssemblies = new Dictionary<string, Assembly>();
+        private readonly List<string> _dependencyPaths = new List<string>();
+        private readonly object _guard = new object();
+
+        private bool _hookedAssemblyResolve = false;
+
+        /// <summary>
+        /// Add the path to an assembly to be considered while handling the
+        /// <see cref="AppDomain.AssemblyResolve"/> event.
+        /// </summary>
+        public void AddDependencyLocation(string fullPath)
+        {
+            if (fullPath == null)
+            {
+                throw new ArgumentNullException(nameof(fullPath));
+            }
+
+            lock (_guard)
+            {
+                if (!_dependencyPaths.Contains(fullPath, StringComparer.OrdinalIgnoreCase))
+                {
+                    _dependencyPaths.Add(fullPath);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Given a path to an assembly on disk, loads and returns the
+        /// corresponding <see cref="Assembly"/> object.
+        /// </summary>
+        /// <remarks>
+        /// Multiple calls with the same path will return the same 
+        /// <see cref="Assembly"/> instance.
+        /// 
+        /// Also hooks the <see cref="AppDomain.AssemblyResolve"/> event in
+        /// case it needs to load a dependency.
+        /// </remarks>
+        public Assembly LoadFromPath(string fullPath)
+        {
+            if (fullPath == null)
+            {
+                throw new ArgumentNullException(nameof(fullPath));
+            }
+
+            Debug.Assert(PathUtilities.IsAbsolute(fullPath));
+
+            lock (_guard)
+            {
+                Assembly assembly;
+                if (_pathsToAssemblies.TryGetValue(fullPath, out assembly))
+                {
+                    return assembly;
+                }
+
+                assembly = LoadInternal(fullPath);
+
+                if (!_hookedAssemblyResolve)
+                {
+                    _hookedAssemblyResolve = true;
+
+                    AppDomain.CurrentDomain.AssemblyResolve += CurrentDomain_AssemblyResolve;
+                }
+
+                return assembly;
+            }
+        }
+
+        /// <summary>
+        /// 
```

Comment:
Internal method, but still good to fix this and move the abstract method near the top of the file?